lmdeploy/pytorch/paging/block_trie.py

```diff
@@ -0,0 +1,216 @@
+            self.leaves.remove(node)
+
+        block_id = num_matched // block_size
+        blocks = []
+        free_blocks = []
+        while num_matched + block_size <= num_all_ids:
+            curr_tokens = seq.history_cache[num_matched:num_matched +
+                                            block_size]
+
+            block = logical_blocks[block_id]
+
+            hash_key = hash(tuple(curr_tokens))
+            parent = node
+            if hash_key in parent.children:
+                child = parent.children[hash_key]
+                if not np.array_equal(curr_tokens, child.tokens):
+                    break
+                node = child
+                free_blocks.append(block)
+                logical_blocks[block_id] = node.block
+            else:
+                node = Node(hash_key=hash_key,
+                            block=block,
+                            tokens=curr_tokens,
+                            num_matched=num_matched + block_size)
+                node.parent = parent
+            blocks.append(node.block)
+            num_matched += block_size
+            block_id += 1
+
+        logical_blocks.last_shared_node = node
+        if node.parent is not None and len(node.children) == 0:
+            # ignore root
+            self.leaves.add(node)
+        if len(blocks) > 0:
+            self.allocator.add_ref_count(np.array(blocks), 1)
+        if len(free_blocks) > 0:
+            self.allocator.free(np.array(free_blocks))
+
+    def evict(self, max_num_blocks: int):
+        """evict."""
+        if not self.enable:
+            return 0
+
+        def __remove_leaf(leaves, evicted_blocks):
+            _, leaf = heapq.heappop(leaves)
+            evicted_blocks.append(leaf.block)
+            parent = leaf.parent
+            leaf.parent = None
+            self.leaves.remove(leaf)
+            return parent
+
+        def __add_leaf(leaves, parent):
+            self.leaves.add(parent)
+            if self.allocator.get_ref_count(parent.block) == 1:
+                access_time = self.allocator.get_access_time(parent.block)
+                heapq.heappush(leaves, (access_time, parent))
+
+        evicted_blocks = []
+        leaves = list(self.leaves)
+
+        # filter ref-cnt == 1 (trie own one block ref)
+        leave_blocks = np.array(list(leaf.block for leaf in leaves))
+        ref_cnt = self.allocator.get_ref_count(leave_blocks)
+        access_times = self.allocator.get_access_time(leave_blocks)
+        indices = (ref_cnt == 1).nonzero()[0]
+        if len(indices) == 0:
+            return 0
+
+        # make heap
+        leaves = list(leaves[i] for i in indices)
+        access_times = list(access_times[i] for i in indices)
+        leaves = list(zip(access_times, leaves))
+        heapq.heapify(leaves)
+
+        while len(leaves) > 0 and len(evicted_blocks) < max_num_blocks:
+            parent = __remove_leaf(leaves, evicted_blocks)
+            if parent.parent is None:
```

Comment:
parent is root, but `len(leaves) > 0` (root may have other children), there are still some blocks can be evicted